klib/ntp.c

```diff
@@ -174,18 +184,34 @@ static void ntp_input(void *z, struct udp_pcb *pcb, struct pbuf *p,
     }
 
     /* If at least 2 samples have been received from the NTP server, calculate a calibration value
-     * to be applied after the local time is synchronized with the NTP time. */
+     * to be applied after the local time is synchronized with the NTP time, and possibly adjust the
+     * query interval based on the clock jitter. */
     raw -= rtd / 2;
     if ((ntp.last_raw != 0) && (raw > ntp.last_raw)) {
         cal = (s64)div128_64(offset_calibr, raw - ntp.last_raw);
         if ((cal < 0) || (cal > NTP_MAX_SLEW_RATE))
             cal = NTP_MAX_SLEW_RATE;
         if (offset < 0)
             cal = -cal;
+        s64 jitter = offset - ntp.last_offset;
+        if (ABS(offset) > ABS(jitter) * 4) {
+            ntp.jiggle_counter += ntp.query_interval;
+            if ((ntp.jiggle_counter > NTP_JIGGLE_THRESHOLD) && (ntp.query_interval < ntp.pollmax)) {
+                ntp.query_interval++;
+                ntp.jiggle_counter = 0;
+            }
+        } else {
+            ntp.jiggle_counter -= 2 * ntp.query_interval;
```

Comment:
Out of curiosity, why is the query_interval factor doubled here but not in the previous block?

Reply:
The rationale here would be that the NTP client show react faster (decreasing the poll interval) when the offset increases, whereas it could react more slowly when things are improving. It's what the code skeleton in the NTP RFC does (https://tools.ietf.org/html/rfc5905, page 102).
By the way, I realized there was a mistake, the `(ABS(offset) > ABS(jitter) * 4)` condition should be `(ABS(offset) < ABS(jitter) * 4)` instead (I was misled by https://www.eecis.udel.edu/~mills/ntp/html/poll.html, which contains a mistake). I corrected this now and force-pushed.